File: 정종길/프로그래머스/산 모양 타일링.py
def solution(n, tops):
    if tops[0] == 1:
        an = 3
        bn = 4
    else:
        an = 2
        bn = 3

    for i, top in enumerate(tops[1:], start=1):
        if top:
            new_bn = 3 * bn + an
            new_an = 2 * bn + an
        else:
            new_bn = 2 * bn + an
            new_an = bn + an

        bn, an = new_bn, new_an

    answer = bn % 10007

    return answer

File: 정종길/프로그래머스/카카오 3회독/2024 KAKAO WINTER INTERNSHIP/산 모양 타일링.java
import java.util.*;

class Solution {
    public int solution(int n, int[] tops) {
        
        int[] an = new int[n];
        int[] bn = new int[n];
        
        if (tops[0] == 1){
            an[0] = 4;
            bn[0] = 3;
        }else{
            an[0] = 3;
            bn[0] = 2;
        }
        
        for(int i = 1; i < n; i++){
            if(tops[i] == 1){
                an[i] = 3 * an[i - 1] + bn[i - 1];
                bn[i] = 2 * an[i - 1] + bn[i - 1];
            }else{
                an[i] = 2 * an[i - 1] + bn[i - 1];
                bn[i] = an[i - 1] + bn[i - 1];
            }
            an[i] %= 10007;
            bn[i] %= 10007;
        }
        return an[n-1];
    }
}